x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchIndexService.java

```diff
@@ -265,14 +265,17 @@ void getResponse(AsyncSearchId searchId,
                     return;
                 }
 
-                if (restoreResponseHeaders) {
+                if (restoreResponseHeaders && get.getSource().containsKey(RESPONSE_HEADERS_FIELD)) {
```

Comment:
this seems at first glance unrelated, did it come up from the tests you added as part of this PR?

Reply:
Yes, this fixes an NPE when retrieving the initial result (if the coordinating node died during the request). We discussed some options here,  whether we should throw a resource not found  or not but this change is just an hot-fix